### PR TITLE
mkosi: make -i mode snappier on reflink-capable file systems

### PR DESCRIPTION
**Pull request by poettering**
Let's start out with a reflink copy, even if we don't actually want CoW
for disk images ultimately.

### CODE REVIEW

File: mkosi [hunk 1]
```diff
@@ -415,8 +415,22 @@ def reuse_cache_image(args, workspace, run_build_script, for_cache):
         with source:
             f = tempfile.NamedTemporaryFile(dir = os.path.dirname(args.output), prefix='.mkosi-')
             output.append(f)
+
+            # So on one hand we want CoW off, since this stuff will
+            # have a lots of random write accesses. On the other we
```
Comment (reverendhomer) [inline mkosi]: Just a little nitpick: typo here. "a lot of" or "lots of" would be right.

**Comment by poettering** [comment]
I force pushed a new version now that fixes the typo @reverendhomer pointed out. PTAL!

File: mkosi [hunk 2]
```diff
@@ -415,8 +415,22 @@ def reuse_cache_image(args, workspace, run_build_script, for_cache):
...
+            # CoW but later changes do not result in CoW anymore.
+
             disable_cow(f.name)
-            shutil.copyfileobj(source, f)
+            try:
+                _reflink(source.fileno(), f.fileno())
+            except OSError as e:
+                if e.errno not in {errno.EXDEV, errno.EOPNOTSUPP}:
+                    raise
+                shutil.copyfileobj(source, f)
```
Comment (lucaswerkmeister) [inline mkosi]: Couldn’t this use the `_copy_file` function? It looks like it does mostly the same thing, and I’m not sure if the slight differences are intended.

File: mkosi [hunk 3]
```diff
@@ -415,8 +415,22 @@ def reuse_cache_image(args, workspace, run_build_script, for_cache):
...
             f = tempfile.NamedTemporaryFile(dir = os.path.dirname(args.output), prefix='.mkosi-')
             output.append(f)
+
+            # So on one hand we want CoW off, since this stuff will
+            # have a lot of random write accesses. On the other we
+            # want the copy to be snappy, hence we do want CoW. Let's
+            # ask for both, and let the kernel figure things out:
+            # let's turn off CoW on the file, but start with a CoW
+            # copy. On btrfs that works: the initial copy is made as
+            # CoW but later changes do not result in CoW anymore.
```
Comment (lucaswerkmeister) [inline mkosi]: I found it a bit confusing that this long block of documentation precedes just a single call to `_reflink` – where is all this “turn off CoW but start with CoW” logic hiding? :) IIUC that’s just the behavior of the `FICLONE` ioctl, so I think some of this documentation should be placed on `_reflink` instead.

Comment (poettering) [reply mkosi]: not quite. we first invoke "disable_cow()" and then follow up with "_reflink()" right after. it's that combination of first turning off cow, and then doing a reflink (which is the mother of all cow concepts) after all that deserves the comment.

File: mkosi (hunk 2 above)
Comment (poettering) [reply mkosi]: copy_file() takes a path, _copy_file() takes fd integers, but we need python file objects here...

But you are right, we should probably split this into a function of its own. Will do.

**Comment by poettering** [comment]
Force pushed a new version now, which turns the copy stuff into a function of its own. please have a look. it's a bit nasty that we currently have so many different slightly different flavours of copy file routines, but cleaning that up is probably work for another PR

File: mkosi (hunk 3 above)
Comment (lucaswerkmeister) [reply mkosi]: Wait, how the fuck did I not see that `disable_cow` call? :D

Ben Mendelsohn playing Orson Krennic in “Rogue One: A Star Wars Story”, in the scene on Scarif just after the rebel attack starts, yelling at the officers surrounding him: “Are we blind? Deploy the garrison!”

**Pull request merged**